### PR TITLE
Home page

### PR DESCRIPTION
**Pull request by zpawn**
Story https://app.shortcut.com/deskpro/story/156161/lansweeper-home-page

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -32,3 +32,23 @@ export const APP_URL = "https://app.lansweeper.com";
...
+    `"assetCustom.model"`,
+    `"assetCustom.serialNumber"`,
+    `"diskPartitions.mountedOn"`,
+    `"diskPartitions.available"`,
+    `"diskPartitions.size"`,
+    `"operatingSystem.caption"`,
+    `"operatingSystem.name"`,
+    `"operatingSystem.version"`,
+  ],
+};
```
Comment (HappyPaul55) [inline src/constants.ts:54]: Can add `as const` at the end for more accurate typings.

File: src/pages/HomePage/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { QueryKey } from "../../query";
+import type { DeviceType, UserContext } from "../../types";
+
+type UseLinkedDevices = () => {
+  isLoading: boolean;
+  devices: DeviceType[];
+};
+
+const useLinkedDevices: UseLinkedDevices = () => {
+  const { context } = useDeskproLatestAppContext() as { context: UserContext };
```
Comment (HappyPaul55) [inline src/pages/HomePage/hooks.ts:19]: We shouldn't be typing here. The SDK should be providing proper types and if not, since we own that SDK, we should fix it.

File: src/services/lansweeper/getDevicesService.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        sites {
+          id
+          name
+          brandingName
+          assetResources(
+            fields: [${fields.DEVICE.join(",")}],
+            filters: {
+              conjunction: OR
+              conditions: [
+                ${deviceKeys.map((key) => `{ operator: EQUAL, path: "key", value: "${key}" }`).join(',')}
```
Comment (HappyPaul55) [inline src/services/lansweeper/getDevicesService.ts:23]: Does there need to be any escaping here?

File: src/hooks/useBadgeCount.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { size } from "lodash";
+import { useInitialisedDeskproAppClient } from "@deskpro/app-sdk";
+import type { DeviceType } from "../types";
+
+const useBadgeCount = (items: DeviceType[]) => {
+  useInitialisedDeskproAppClient((client) => {
+    if (!Array.isArray(items)) {
```
Comment (HappyPaul55) [inline src/hooks/useBadgeCount.ts:7]: Items is always an array so this check isn't required.

File: src/hooks/useBadgeCount.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      return;
+    }
+
+    client.setBadgeCount(size(items));
```
Comment (HappyPaul55) [inline src/hooks/useBadgeCount.ts:11]: This can be `client.setBadgeCount(items.length);`. No need for 3rd party libraries as we're dealing with native/normal arrays.

**Review by HappyPaul55 — CHANGES_REQUESTED** [review]

File: src/pages/HomePage/hooks.ts (hunk 2 above)
Comment (zpawn) [reply src/pages/HomePage/hooks.ts:19]: I created a story to fix it in SDK https://app.shortcut.com/deskpro/story/158090/app-sdk-should-fix-context-typings

File: src/services/lansweeper/getDevicesService.ts (hunk 3 above)
Comment (zpawn) [reply src/services/lansweeper/getDevicesService.ts:23]: No, not needed. We get `deviceKeys` from our backend

File: src/hooks/useBadgeCount.ts (hunk 4 above)
Comment (zpawn) [reply src/hooks/useBadgeCount.ts:7]: Fixed typing. I'm not sure it's always an array.

Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts:7]: I don't think that's the correct typing, `useBadgeCount` is always called with one argument, so it will always be an `array`.

Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts:7]: If you meant `(items?: DeviceType[]|undefined)` then I don't think that's a good signature and the function has unexpected behaviour.

If `undefined` is sent, why doesn't it set the count to `0`. Instead I strongly recommend we keep this simple and it always passes in an array and we just always call `client.setBadgeCount(...)`

File: src/pages/HomePage/hooks.ts (hunk 2 above)
Comment (HappyPaul55) [reply src/pages/HomePage/hooks.ts:19]: Cool beans, but can we add more to the story. A single screenie doesn't say much as is missing the larger picture. Whilst we're here, might as well say why you typed it what you expected the SDK to do instead.

File: src/hooks/useBadgeCount.ts (hunk 4 above)
Comment (zpawn) [reply src/hooks/useBadgeCount.ts]: I change the function like this. The `items` come from the API, so I'm not sure if the array will come

Screenshot 2024-07-18 at 13 17 02

File: src/hooks/useBadgeCount.ts (hunk 5 above)
Comment (zpawn) [reply src/hooks/useBadgeCount.ts]: We have an approach to use `lodash`. If you don't want to use `lodash`, don't use `lodash` in your apps

File: src/pages/HomePage/hooks.ts (hunk 2 above)
Comment (zpawn) [reply src/pages/HomePage/hooks.ts:19]: Added more details

File: src/hooks/useBadgeCount.ts (hunk 5 above)
Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts:7]: Where is that documented/agreed? Maybe one for a retro because using `lodash` everywhere when it adds no value of native functions seems like a questionable life choice.

Comment (zpawn) [reply src/hooks/useBadgeCount.ts:7]: It's a good topic for a retro discussion.

Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts:7]: As discussed in Slack chat, I recommend we change this whilst it's fresh in our heads. lodash has no value over `.length` here so we should swap it out.

That said, it's probably a good idea to get a 3rd party to have a fresh lookie at it. Thoughts @AshleyDawson

Comment (AshleyDawson) [reply src/hooks/useBadgeCount.ts:7]: My opinion is that we should probably lean on core language features over libraries, in this case I think `array.length` is more appropriate. That said, lodash can be useful for some things, and in this case it's important to only only add the specific lodash function as I'm not sure that we currently tree-shake correctly.

We've leaned on lodash in the past as it's been convenient if already installed - but I strongly suggest that we don't include the entire lodash library going forward... the only thing it will do is increase the bundle size.

tldr; I think best practice should be: stick to core language features as much as possible and include libraries where absolutely necessary.

Comment (zpawn) [reply src/hooks/useBadgeCount.ts:7]: > We've leaned on lodash in the past as it's been convenient if already installed - but I strongly suggest that we don't include the entire lodash library going forward... the only thing it will do is increase the bundle size.

I checked, and using individual `lodash` modules like `lodash.size`, `lodash.debounce`, etc., includes all the internal helper functions of `lodash`. If multiple individual modules are installed, they end up being larger than the standalone `lodash` package. 

At that time with David P., we decided that if you are using only one `lodash` function, it's better to install the individual module. If you are using two or more functions, then it's best to install the whole `lodash` package.

Comment (zpawn) [reply src/hooks/useBadgeCount.ts]: changed to `client.setBadgeCount(items?.length ?? 0);`

Comment (AshleyDawson) [reply src/hooks/useBadgeCount.ts]: @zpawn I propose that we always install individual lodash functions

Comment (zpawn) [reply src/hooks/useBadgeCount.ts]: @AshleyDawson if we use one function - better install one module. But if we use more than one function - better use `lodash` package

Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts]: Personally, I don't think we need any of `lodash` from what I've seen. `get(a, ["b","c"])` can just be `a?.b?.c;`, `map` is already in JS/TS and as we're mentioning here, `size` can be replaced `foo.length`. To me, it looks like `lodash` is an extra thing to learn which replaces/reimplements native features from the language itself but is another vector for supply chain attacks and more bloating/loading.

With proper typings and linting (e.g. `@typescript-eslint/strict-type-checked`) we probably have everything we need to build safe, performant software efficiently.

Comment (zpawn) [reply src/hooks/useBadgeCount.ts]: @HappyPaul55 it doesn’t seem to be a critical moment. It would be great if we tried to keep the balance between criticism and wasting time.

File: src/components/DeviceItem/DeviceItem.tsx [hunk 6]
```diff
@@ -1,61 +1,58 @@
...
 import type { FC } from "react";
-import type { Maybe } from "../../types";
-import type { Site, Device } from "../../services/lansweeper/types";
+import type { Maybe, DeviceType } from "../../types";
 
 export type Props = {
-  device: Device;
-  sites: Site[];
-  siteId: Maybe<Site["id"]>;
+  device: Maybe<DeviceType>;
```
Comment (HappyPaul55) [inline src/components/DeviceItem/DeviceItem.tsx:10]: There is no need for this to be `Maybe`. Following the code that calls this, `device` is always set to `DeviceType` so we don't need to complicate this function by catering for things that don't happen.

source: https://github.com/DeskproApps/lansweeper/pull/10/files#diff-ec3c1fdc70706cb1636369651ddafe7432d9d1b2ca05263e75ca0fc1d04fc828R20
source: https://github.com/DeskproApps/lansweeper/pull/10/files#diff-147dc47545e2ab3bc81f3899c5a12fd2b02a87d9f655413e88e542229900310aR51

File: src/components/Home/Home.tsx [hunk 7]
```diff
@@ -1,13 +1,27 @@
...
 };
 
-const Home: FC<Props> = () => {
+const Home: FC<Props> = ({ devices }) => {
   return (
-    <Title title="Home Page"/>
+    <Container>
+      {isEmpty(devices) ? (
+        <NotFound text="No devices found" />
+      ) : map(devices, (device) => (
```
Comment (HappyPaul55) [inline src/components/Home/Home.tsx:18]: Replace `map` with JS/TS native `map` - no need for lodash.

File: src/components/Home/Home.tsx [hunk 8]
```diff
@@ -1,13 +1,27 @@
...
+export type Props = {
+  devices: DeviceType[];
 };
 
-const Home: FC<Props> = () => {
+const Home: FC<Props> = ({ devices }) => {
   return (
-    <Title title="Home Page"/>
+    <Container>
+      {isEmpty(devices) ? (
```
Comment (HappyPaul55) [inline src/components/Home/Home.tsx:16]: Replace `isEmpty` with JS/TS native `.length === 0` - no need for `lodash`.

File: src/pages/HomePage/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  const devices = useQueryWithClient(
+    [QueryKey.DEVICES, ...(Array.isArray(linkedIds.data) ? linkedIds.data : []) as string[]],
+    (client) => getDevicesService(client, linkedIds.data as string[]),
+    { enabled: (Array.isArray(linkedIds.data) ? linkedIds.data : []).length > 0 },
+  );
+
+  return {
+    isLoading: [linkedIds, devices].some(({ isLoading }) => isLoading),
+    devices: useMemo(() => enhanceDevices(get(devices.data, ["data"])), [devices.data]),
```
Comment (HappyPaul55) [inline src/pages/HomePage/hooks.ts:36]: Replace `get` with JS/TS native `foo?.bar`- no need for lodash.

File: src/pages/LinkDevicePage/LinkDevicePage.tsx [hunk 10]
```diff
@@ -24,15 +24,15 @@ const LinkDevicePage: FC = () => {
   const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
   const [siteId, setSiteId] = useState<Maybe<Site["id"]>>(null);
   const [searchQuery, setSearchQuery] = useState<string>("");
-  const [selectedDevices, setSelectedDevices] = useState<Device[]>([]);
+  const [selectedDevices, setSelectedDevices] = useState<DeviceType[]>([]);
   const { isLoading, isFetching, sites, devices } = useSearchDevices(siteId, searchQuery);
-  const dpUserId = useMemo(() => get(context, ["data", "user", "id"]), [context]);
+  const dpUserId = useMemo(() => context?.data?.user.id, [context]);
```
Comment (HappyPaul55) [inline src/pages/LinkDevicePage/LinkDevicePage.tsx:29]: >Unnecessary optional chain on a non-nullish value
`context` is always an `object`

File: src/components/LinkDevice/blocks/Filters.tsx [hunk 11]
```diff
@@ -15,7 +15,7 @@ type Props = {
 };
 
 const Filters: FC<Props> = ({ sites, siteId, isFetching, onChangeSearchQuery, onChangeSite }) => {
-  const options = useMemo(() => getOptions(sites, "companyName"), [sites]);
+  const options = useMemo(() => getOptions(sites, "brandingName"), [sites]) as Array<Option>;
```
Comment (HappyPaul55) [inline src/components/LinkDevice/blocks/Filters.tsx:18]: Can this be typed without using `as`? My guess is `getOptions` is missing some type information which is why it needs to be set it.

File: src/pages/HomePage/hooks.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import type { DPUser, DeviceType, UserContext } from "../../types";
+
+type UseLinkedDevices = () => {
+  isLoading: boolean;
+  devices: DeviceType[];
+};
+
+const useLinkedDevices: UseLinkedDevices = () => {
+  const { context } = useDeskproLatestAppContext() as { context: UserContext };
+  const dpUserId = get(context, ["data", "user", "id"]);
```
Comment (HappyPaul55) [inline src/pages/HomePage/hooks.ts:20]: Replace `get` with JS/TS native `foo?.bar`- no need for lodash.

File: src/pages/LinkDevicePage/hooks.ts [hunk 13]
```diff
@@ -20,21 +21,27 @@ const getSites = (profiles?: Profile[]) => {
...
+  }, [sitesResponce.data]);
+
+  const devicesResponce = useQueryWithClient(
     [QueryKey.DEVICES, siteId as Site["id"], q as string],
     (client) => searchDevicesService(client, { siteId: siteId as Site["id"], q: q as string }),
     { enabled: Boolean(q) && Boolean(siteId) },
   );
 
+  const devices: DeviceType[] = useMemo(() => {
+    return enhanceSearchDevices(find(sites, { id: siteId }) as Site, devicesResponce.data?.data);
```
Comment (HappyPaul55) [inline src/pages/LinkDevicePage/hooks.ts:37]: Replace `find` with JS/TS native `.find` - no need for lodash.

Comment (HappyPaul55) [inline src/pages/LinkDevicePage/hooks.ts:37]: If I've understhood this correctly, `enhanceSearchDevices` returns `DeviceType[]` but it typed to `Site`. Is this not a type mismatch?

File: src/utils/enhanceDevice.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+const enhanceDevices = (data?: DevicesResponse): DeviceType[] => {
+  const sites = data?.authorizedSites?.sites ?? [];
+
+  if (!sites) {
+    return [];
+  }
+
+  return reduce(sites, (acc: DeviceType[], s) => {
```
Comment (HappyPaul55) [inline src/utils/enhanceDevice.ts:27]: Replace `reduce` with JS/TS native `reduce` - no need for lodash.

File: src/utils/enhanceDevice.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  if (isEmpty(site) || isEmpty(data) || !Array.isArray(items)) {
+    return [];
+  }
+
+  return items.map((d: Device) => {
+    return assign({}, d, { site }) as DeviceType;
+  });
+};
+
+const enhanceDevices = (data?: DevicesResponse): DeviceType[] => {
```
Comment (HappyPaul55) [inline src/utils/enhanceDevice.ts:20]: This is probably meant to read `(data: DevicesResponse|undefined)` as `data` must always be sent in but it can be undefined. 

Personally I'd removed that complexity and instead just say that `data` must always be of type `DevicesResponse` and enforce the calling code sending in the right data.

File: src/utils/parseGlobalAccessToken.ts [hunk 16]
```diff
@@ -15,8 +14,8 @@ const parseGlobalAccessToken: ParseGlobalAccessToken = (tokens) => {
 
   try {
     const parsed = JSON.parse(tokens);
```
Comment (HappyPaul55) [inline src/utils/parseGlobalAccessToken.ts:16]: Should probably add a type on this so we know the expected result.

**Review by HappyPaul55 — CHANGES_REQUESTED** [review]

File: src/hooks/useBadgeCount.ts [hunk 17]
```diff
@@ -0,0 +1,10 @@
+import { useInitialisedDeskproAppClient } from "@deskpro/app-sdk";
+import type { Maybe, DeviceType } from "../types";
+
+const useBadgeCount = (items: Maybe<DeviceType[]>) => {
```
Comment (HappyPaul55) [inline src/hooks/useBadgeCount.ts:4]: Type needs updating to `any[]` as it doesn't care about what is contained in the array. Or maybe use `count: number` instead. Eitherway, it doesn't need `Maybe` as it is and must always be called with an `array`, there is no reason to call this without an entry.

File: src/components/DeviceItem/DeviceItem.tsx (hunk 6 above)
Comment (zpawn) [reply src/components/DeviceItem/DeviceItem.tsx:10]: done

File: src/components/Home/Home.tsx (hunk 7 above)
Comment (zpawn) [reply src/components/Home/Home.tsx:18]: done

File: src/pages/HomePage/hooks.ts (hunk 9 above)
Comment (zpawn) [reply src/pages/HomePage/hooks.ts:36]: done

File: src/pages/LinkDevicePage/LinkDevicePage.tsx (hunk 10 above)
Comment (zpawn) [reply src/pages/LinkDevicePage/LinkDevicePage.tsx:29]: I'm afraid context doesn't always is object.
Screenshot 2024-07-24 at 16 28 53

Comment (zpawn) [reply src/pages/LinkDevicePage/LinkDevicePage.tsx:29]: Here story for fix types https://app.shortcut.com/deskpro/story/158090/app-sdk-should-fix-context-typings

File: src/components/LinkDevice/blocks/Filters.tsx (hunk 11 above)
Comment (zpawn) [reply src/components/LinkDevice/blocks/Filters.tsx:18]: I'm afraid not. Need to fix types in "app-sdk". I created story for it https://app.shortcut.com/deskpro/story/158765/app-sdk-need-to-fix-the-types-in-select-component

File: src/pages/LinkDevicePage/hooks.ts (hunk 13 above)
Comment (zpawn) [reply src/pages/LinkDevicePage/hooks.ts:37]: You understand correctly because the assets are located on the site.

File: src/pages/HomePage/hooks.ts (hunk 12 above)
Comment (zpawn) [reply src/pages/HomePage/hooks.ts:20]: done

File: src/utils/enhanceDevice.ts (hunk 15 above)
Comment (zpawn) [reply src/utils/enhanceDevice.ts:20]: Understood your point. Changed to `(data: Maybe<SearchResponce>)`. Because raw data from the API response comes here.

File: src/utils/parseGlobalAccessToken.ts (hunk 16 above)
Comment (zpawn) [reply src/utils/parseGlobalAccessToken.ts:16]: already added
Screenshot 2024-07-24 at 17 20 38

File: src/hooks/useBadgeCount.ts (hunk 17 above)
Comment (zpawn) [reply src/hooks/useBadgeCount.ts:4]: There are two points here:
1. We have a rule - not to use the type any.
2. Based on the application context, only `DeviceType[]` can be used here.
3. We have already discussed this hook. And I implemented it as we agreed

File: src/components/LinkDevice/blocks/Filters.tsx (hunk 11 above)
Comment (HappyPaul55) [reply src/components/LinkDevice/blocks/Filters.tsx:18]: `getOptions` is in this codebase, not the SDK, and so it `Option`. I'd expect `getOptions` to return `Option[]` so it doesn't have to be typed here.

Comment (zpawn) [reply src/components/LinkDevice/blocks/Filters.tsx:18]: Could you help me with this type, please? How it must be

File: src/hooks/useBadgeCount.ts (hunk 17 above)
Comment (HappyPaul55) [reply src/hooks/useBadgeCount.ts:4]: 1. To avoid `any`, use `count: number`.
2. We should follow separation of concern, the `useBadgeCount` doesn't need to know anything about devices or `DeviceType`. It's coupling things that don't need need to be coupled.
3. We discussed in Slack but you implemented something else that was not agreed.

File: src/utils/parseGlobalAccessToken.ts (hunk 16 above)
Comment (HappyPaul55) [reply src/utils/parseGlobalAccessToken.ts:16]: I meant as `const parsed = JSON.parse(tokens) as {access_token: string,...};`

File: src/components/DeviceItem/DeviceItem.tsx [hunk 18]
```diff
@@ -1,61 +1,58 @@
...
-  siteId: Maybe<Site["id"]>;
+  device: DeviceType;
   onClickTitle?: () => void,
 };
 
-const DeviceItem: FC<Props> = ({ sites, siteId, device, onClickTitle }) => {
+const DeviceItem: FC<Props> = ({ device, onClickTitle }) => {
   const { getDeviceLink } = useExternalLinks();
-  const site = useMemo(() => find(sites, { id: siteId }), [sites, siteId]) as Maybe<Site>;
+  const site = device?.site;
```
Comment (HappyPaul55) [inline src/components/DeviceItem/DeviceItem.tsx:16]: `device` is always an object based on typings, so no need for `?.`. Either change code or type.

File: src/hooks/useBadgeCount.ts (hunk 17 above)
Comment (zpawn) [reply src/hooks/useBadgeCount.ts:4]: This hook calculates the number of assets. So everything is correct here with concern

File: src/components/DeviceItem/DeviceItem.tsx (hunk 18 above)
Comment (zpawn) [reply src/components/DeviceItem/DeviceItem.tsx:16]: Yesterday in Slack you said the opposite

File: src/utils/parseGlobalAccessToken.ts (hunk 16 above)
Comment (zpawn) [reply src/utils/parseGlobalAccessToken.ts:16]: I don't see the need because the function is typed, and we can see what it returns.

File: src/components/DeviceItem/DeviceItem.tsx (hunk 18 above)
Comment (HappyPaul55) [reply src/components/DeviceItem/DeviceItem.tsx:16]: You may have changed the typings. I was using the output from TS.

**Review by HappyPaul55 — APPROVED** [review]
